### PR TITLE
Refactor the usage of batch catfile

### PR DESCRIPTION
**Pull request by lunny**
When opening a repository,  it will call `ensureValidRepository` and also `CatFileBatch`. But sometimes these will not be used until repository closed. So it's a waste of CPU to invoke 3 times git command for every open repository.

This PR removed all of these from `OpenRepository` but only kept checking whether the folder exists. When a batch is necessary, the necessary functions will be invoked.

### CODE REVIEW

**Review by wolfogre — APPROVED** [review]

File: modules/indexer/code/elasticsearch/elasticsearch.go [hunk 1]
```diff
@@ -154,25 +154,27 @@ func (b *Indexer) addDelete(filename string, repo *repo_model.Repository) elasti
...
+			updateReqs, err := b.addUpdate(ctx, batch.Writer, batch.Reader, sha, update, repo)
 			if err != nil {
 				return err
 			}
 			if len(updateReqs) > 0 {
 				reqs = append(reqs, updateReqs...)
 			}
 		}
-		cancel()
+		batch.Close()
```
Comment (yp05327) [inline modules/indexer/code/elasticsearch/elasticsearch.go:177]: L166 defer batch.Close()
L177 batch.Close()

Do we need this twice? 🤔

Comment (lunny) [reply modules/indexer/code/elasticsearch/elasticsearch.go:177]: Yes. L166 ensure close when `return err` invoked before L177. And it's safe to invoke twice of `batch.Close`

File: modules/git/repo_base_nogogit.go [hunk 2]
```diff
@@ -55,63 +51,75 @@ func OpenRepository(ctx context.Context, repoPath string) (*Repository, error) {
...
+		repo.batch, err = repo.NewBatch(ctx)
+		if err != nil {
+			return nil, nil, nil, err
+		}
 	}
-	repo.batchInUse = true
-	return repo.batchWriter, repo.batchReader, func() {
-		repo.batchInUse = false
+
+	if !repo.batchInUse {
```
Comment (yp05327) [inline modules/git/repo_base_nogogit.go:71]: Just a question.
Why `repo.batch` should always not be `nil`? It is a bit different to the old logic.

Comment (lunny) [reply modules/git/repo_base_nogogit.go:71]: I don't think so. The original logic, all of them will be init when open the repository. The new logic is a lazy loading. `repo.batch` will be init for the first invoking `CatFileBatch`.

**Review by yp05327 — APPROVED** [review]
I think it is better to adjust the newly added logs, they are copy and paste, same logs from different places are not good for debug.

**Comment by lunny** [comment]
> I think it is better to adjust the newly added logs, they are copy and paste, same logs from different places are not good for debug.

I will try to merge the two functions in following PRs

**Pull request merged**